src/callbacks/memory_profiler.cpp

```diff
@@ -0,0 +1,578 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2023, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+//
+// memory_profiler .hpp .cpp - Itemized memory usage profiling.
+///////////////////////////////////////////////////////////////////////////////
+
+#include "lbann/callbacks/memory_profiler.hpp"
+#include "lbann/layers/data_type_layer.hpp"
+#include "lbann/models/model.hpp"
+#include "lbann/utils/gpu/helpers.hpp"
+#include "lbann/utils/serialize.hpp"
+#include "lbann/weights/data_type_weights.hpp"
+#include "lbann/weights/weights.hpp"
+
+#include "lbann/proto/callbacks.pb.h"
+
+#include <algorithm>
+#include <string>
+
+namespace lbann {
+namespace callback {
+
+/**
+ * @brief Returns the currently used memory, or 0 if LBANN was not compiled with
+ * GPU support.
+ */
+static inline size_t get_used_gpu_memory()
+{
+#ifdef LBANN_HAS_GPU
+  size_t available;
+  size_t total;
+#ifdef LBANN_HAS_CUDA
+  FORCE_CHECK_CUDA(cudaMemGetInfo(&available, &total));
+#elif defined(LBANN_HAS_ROCM)
+  FORCE_CHECK_ROCM(hipMemGetInfo(&available, &total));
+#endif
```

Comment:
```suggestion
auto const [available, total] = h2::gpu::mem_info();
```
See [H2 GPU memory helpers](https://github.com/LLNL/DiHydrogen/blob/develop/include/h2/gpu/memory_utils.hpp). I expect these to be stable.